<!--
    amcl launch
    Copyright Â© 2018 Naoki Akai. All rights reserved.
 -->

<launch >
    <!-- map -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/nic_garage/ogm.yaml" /> -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/mate1f/ogm.yaml" /> -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)../../maps/ib_minami_3f/ogm.yaml" /> -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/nic6f/ogm.yaml" /> -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/tctsls/ogm.yaml" /> -->
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/e2bs4f/ogm.yaml" /> -->
    <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/nic1f_2107/ogm.yaml" />
    <!-- <arg name="map_yaml_file" default="$(find robot_sim)/../../maps/ib1f/ogm.yaml" /> -->

    <!-- arguments -->
    <arg name="map_frame" default="/map" />
    <arg name="laser_frame" default="/laser" />
    <arg name="base_link_frame" default="/base_link" />
    <arg name="map_topic_name" default="/amcl_map" />
    <arg name="input_odom_topic_name" default="/ypspur_ros/odom" />
    <arg name="input_scan_topic_name" default="/scan" />
    <arg name="min_particle_num" default="200" />
    <arg name="max_particle_num" default="200" />
    <arg name="resample_threshold" default="0.5" />
    <arg name="scan_step" default="10" />
    <arg name="range_reso" default="0.1" />
    <arg name="max_dist_to_obstacle" default="0.5" />
    <arg name="alpha_slow" default="0.0001" />
    <arg name="alpha_fast" default="0.1" />
    <arg name="update_dist" default="0.2" />
    <arg name="update_yaw" default="2.0" />
    <arg name="update_time" default="10000000.0" />
    <arg name="odom_noise_dist_dist" default="0.9" />
    <arg name="odom_noise_head_dist" default="0.5" />
    <arg name="odom_noise_dist_head" default="0.5" />
    <arg name="odom_noise_head_head" default="1.5" />
    <arg name="use_omni_odom" default="false" />
    <arg name="start_x" default="0.0" />
    <arg name="start_y" default="0.0" />
    <arg name="start_yaw" default="0.0" />
    <arg name="initial_cov_xx" default="0.2" />
    <arg name="initial_cov_yy" default="0.2" />
    <arg name="initial_cov_yawyaw" default="1.0" />
    <arg name="pose_publish_hz" default="20.0" />
    <arg name="use_kld_sampling" default="false" />
    <arg name="add_random_particle" default="true" />
    <arg name="add_random_particle_rate" default="0.1" />
    <arg name="dynamic_scan_point_threshold" default="0.9" />
    <arg name="use_nav_core_server" default="false" />
    <arg name="plot_likelihood_distribution" default="false" />
    <arg name="z_hit" default="0.9" />
    <arg name="z_short" default="0.1" />
    <arg name="z_max" default="0.05" />
    <arg name="z_rand" default="0.05" />
    <arg name="z_hit_var" default="0.08" />
    <arg name="lambda_short" default="0.05" />
    <arg name="max_dist_prob" default="0.043937" />
    <arg name="z_rand_mult" default="0.0033333" />
    <arg name="use_beam_model" default="false" />
    <arg name="use_test_range_measurement" default="false" />
    <arg name="use_class_conditional_observation_model" default="false" />

    <!-- tf -->
    <!-- <node name="world_to_map" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 world map 100" /> -->
    <node name="amcl_frame_to_base_link" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 amcl_frame $(arg base_link_frame) 100" />
    <node name="base_link_to_laser" pkg="tf" type="static_transform_publisher" args="0 0 0 0 0 0 $(arg base_link_frame) $(arg laser_frame) 100" />

    <!-- map server -->
    <node name="amcl_map_server" pkg="map_server" type="map_server" args="$(arg map_yaml_file)" unless="$(arg use_nav_core_server)">
        <remap from="/map" to="/amcl_map" />
        <remap from="/map_metadata" to="/amcl_map_metadata" />
    </node >

    <!-- amcl localizer -->
    <node name="amcl" pkg="localizer" type="amcl" output="screen" >
        <param name="/map_frame" value="$(arg map_frame)" type="string" />
        <param name="/laser_frame" value="$(arg laser_frame)" type="string" />
        <param name="/base_link_frame" value="$(arg base_link_frame)" type="string" />
        <param name="/input_map_topic_name" value="$(arg map_topic_name)" type="string" />
        <param name="/input_odom_topic_name" value="$(arg input_odom_topic_name)" type="string" />
        <param name="/input_scan_topic_name" value="$(arg input_scan_topic_name)" type="string" />
        <param name="/min_particle_num" value="$(arg min_particle_num)" type="int" />
        <param name="/max_particle_num" value="$(arg max_particle_num)" type="int" />
        <param name="/resample_threshold" value="$(arg resample_threshold)" type="double" />
        <param name="/scan_step" value="$(arg scan_step)" type="int" />
        <param name="/range_reso" value="$(arg range_reso)" type="double" />
        <param name="/max_dist_to_obstacle" value="$(arg max_dist_to_obstacle)" type="double" />
        <param name="/alpha_slow" value="$(arg alpha_slow)" type="double" />
        <param name="/alpha_fast" value="$(arg alpha_fast)" type="double" />
        <param name="/update_dist" value="$(arg update_dist)" type="double" />
        <param name="/update_yaw" value="$(arg update_yaw)" type="double" />
        <param name="/update_time" value="$(arg update_time)" type="double" />
        <param name="/odom_noise_dist_dist" value="$(arg odom_noise_dist_dist)" type="double" />
        <param name="/odom_noise_head_dist" value="$(arg odom_noise_head_dist)" type="double" />
        <param name="/odom_noise_dist_head" value="$(arg odom_noise_dist_head)" type="double" />
        <param name="/odom_noise_head_head" value="$(arg odom_noise_head_head)" type="double" />
        <param name="/use_omni_odom" value="$(arg use_omni_odom)" type="bool" />
        <param name="/start_x" value="$(arg start_x)" type="double" />
        <param name="/start_y" value="$(arg start_y)" type="double" />
        <param name="/start_yaw" value="$(arg start_yaw)" type="double" />
        <param name="/initial_cov_xx" value="$(arg initial_cov_xx)" type="double" />
        <param name="/initial_cov_yy" value="$(arg initial_cov_yy)" type="double" />
        <param name="/initial_cov_yawyaw" value="$(arg initial_cov_yawyaw)" type="double" />
        <param name="/pose_publish_hz" value="$(arg pose_publish_hz)" type="double" />
        <param name="/use_kld_sampling" value="$(arg use_kld_sampling)" type="bool" />
        <param name="/add_random_particle" value="$(arg add_random_particle)" type="bool" />
        <param name="/add_random_particle_rate" value="$(arg add_random_particle_rate)" type="double" />
        <param name="/use_test_range_measurement" value="$(arg use_test_range_measurement)" type="bool" />
        <param name="/dynamic_scan_point_threshold" value="$(arg dynamic_scan_point_threshold)" type="double" />
        <param name="/use_nav_core_server" value="$(arg use_nav_core_server)" type="bool" />
        <param name="/plot_likelihood_distribution" value="$(arg plot_likelihood_distribution)" type="bool" />
        <param name="/z_hit" value="$(arg z_hit)" type="double" />
        <param name="/z_short" value="$(arg z_short)" type="double" />
        <param name="/z_max" value="$(arg z_max)" type="double" />
        <param name="/z_rand" value="$(arg z_rand)" type="double" />
        <param name="/z_hit_var" value="$(arg z_hit_var)" type="double" />
        <param name="/lambda_short" value="$(arg lambda_short)" type="double" />
        <param name="/max_dist_prob" value="$(arg max_dist_prob)" type="double" />
        <param name="/z_rand_mult" value="$(arg z_rand_mult)" type="double" />
        <param name="/use_beam_model" value="$(arg use_beam_model)" type="bool" />
        <param name="/use_class_conditional_observation_model" value="$(arg use_class_conditional_observation_model)" type="bool" />
    </node >
</launch >
